src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -115,105 +113,147 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
+                using var _5 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var newProjectIdToStateChecksums);
+
+                foreach (var (projectId, oldChecksum) in oldProjectIdToChecksum)
                 {
-                    if (!newMap.ContainsKey(projectId))
-                    {
-                        // we have a project removed
-                        solution = solution.RemoveProject(projectId);
-                    }
+                    var oldProjectState = solutionState.GetRequiredProjectState(projectId);
+
+                    // this should be cheap since we already computed oldSolutionChecksums (which calls into this).
+                    var oldProjectStateChecksums = await oldProjectState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+                    Contract.ThrowIfTrue(oldChecksum != oldProjectStateChecksums.Checksum);
+                    oldProjectIdToStateChecksums.Add(projectId, oldProjectStateChecksums);
                 }
 
-                // changed project
-                foreach (var (projectId, newProjectChecksums) in newMap)
-                {
-                    if (!oldMap.TryGetValue(projectId, out var oldProjectChecksums))
-                    {
-                        continue;
-                    }
+                // sync over the *info* about all the added/changed projects.  We'll want the info so we can determine
+                // what actually changed.
+                using var _6 = PooledHashSet<Checksum>.GetInstance(out var newChecksumsToSync);
+                newChecksumsToSync.AddRange(newProjectIdToChecksum.Values);
 
-                    Contract.ThrowIfTrue(oldProjectChecksums.Checksum == newProjectChecksums.Checksum);
+                var newProjectStateChecksums = await _assetProvider.GetAssetsAsync<ProjectStateChecksums>(
+                    assetHint: AssetHint.None, newChecksumsToSync, cancellationToken).ConfigureAwait(false);
 
-                    solution = await UpdateProjectAsync(solution.GetRequiredProject(projectId), oldProjectChecksums, newProjectChecksums, cancellationToken).ConfigureAwait(false);
+                foreach (var (checksum, newProjectStateChecksum) in newProjectStateChecksums)
+                {
+                    Contract.ThrowIfTrue(checksum != newProjectStateChecksum.Checksum);
+                    var projectId = newProjectStateChecksum.ProjectId;
+                    newProjectIdToStateChecksums.Add(projectId, newProjectStateChecksum);
                 }
 
+                // Now that we've collected the old and new project state checksums, we can actually process them to
+                // determine what to remove, what to add, and what to change.
+                solution = await UpdateProjectsAsync(
+                    solution, isConeSync, oldProjectIdToStateChecksums, newProjectIdToStateChecksums, cancellationToken).ConfigureAwait(false);
```

Comment:
i extracted the actual mutation code into its own method.  It made things cleaner.  That method also only uses the *state* checksums, not the checksums.  So extracting things out made it so that that code didn't access the wrong maps.

Reply:
If you wanted you might be able to take the code up to this point and extract that to a function which essentially becomes "GetTheProjectIdsToSync" which returns oldProjectIdToChecksum and newProjectIdToChecksum. Then it removes the risk of touching some of the earlier maps.

Reply:
(I'm 100% OK with this being merged as is and treating any potential refactoring as a follow up!)